File: problem77.cpp
#include <vector>
#include <iostream>
using namespace std;

class Solution
{
public:
    vector<vector<int>> comb;
    vector<int> current;
    void backtrack(int index, int num, int k)
    {
        if (current.size() == k)
        {
            comb.push_back(current);
            return;
        }

        for (int i = index; i <= num; i++)
        {
            current.push_back(i);
            backtrack(i + 1, num, k);
            current.pop_back();
        }
    }
    vector<vector<int>> combine(int n, int k)
    {
        current.clear();
        comb.clear();
        backtrack(1, n, k);
        return comb;
    }
};

File: problem77.py
class Solution:
    def combine(self, n: int, k: int):
        # two numbers: all possible comb of k from range in n
        # k  = 3 and n = 2; k = [1,1]; [2,1]; [3,1]; [4,1]; ...
        # all comb means that it is backtracking 
        # n = [ 1, 2,3, 4, ...]

        # make the array of numbers from the n first 
        comb = []
        def backtrack(index ,path): # add something here 
            if len(path) == k:
                comb.append(path[:]) # add to combinations 
                return #return null
            
            for i in range(index, n+1):
                path.append(i)
                backtrack(i+1, path)
                path.pop() # pop from the top/latest item from the path

        backtrack(1, [])
        return comb    
